Change cache key definition in environment

pgjones opened the pull request:
In 6671b973e6de5abc46829a27fd3bbb989d68ca3a the load_template method
was altered to use a cache key other than the template name. The key
chosen was the abs path as returned from the loader get_source
method. Unless there is no path in which case the name is
used. Unfortunately this introduced a performance regression, #485, as
the get_source method (in the FileStoreLoader) loads the template
(causing IO).

The purpose of #332 was to allow the loader to change whilst ensuring
the correct template was loaded, i.e. to fix this case

    env.loader = loader1
    env.get_template('index.html') # return loader1/index.html
    env.loader = loader2
    env.get_template('index.html') # also return loader1/index.html because of cache

This commit changes the cache key to be a combination of a hash of the
loader and the template name. Therefore fixing the above case without
calling the get_source method and thereby avoiding the IO load.

This additionally fixes the above use case for loaders that do not return filenames, e.g. `DictLoader`. However it will likely introduce cache misses should properties of the loader change (and hence the hash changes).

jeffwidman commented:
Thanks for putting this together, it's definitely something that needs to get fixed. Your suggested solution makes sense to me.

Looks like some tests are failing though?

pgjones commented:
Thanks, I've rebased with updates to the tests. I've both fixed the braking test and added one to ensure that the correct template is returned after changing the loader.

iceboy-sjtu commented on jinja2/environment.py:
```diff
@@ -769,15 +769,7 @@ def join_path(self, template, parent):
-        # if template is not file, use name for cache key
-        if cache_key is None:
-            cache_key = name
+        cache_key = "%d%s" % (hash(self.loader), name)
```
What about `cache_key = (self.loader, name)`?

This avoids string conversion and hash collision.

pgjones replied:
Good idea, I've updated.

pgjones commented:
I've just rebased following @iceboy-sjtu comment.

pgjones commented:
@jeffwidman Anything else I need to do?

jeffwidman commented:
LGTM, but I'd like another maintainer to take a look as well before merging just in case I missed any edge cases. Especially since the last time someone tried to improve this section of code it introduced this bug... Definitely would like this to land in the next release though.

cc @davidism @untitaker

untitaker commented:
With this we would break cases where somebody sets `env.cache` to something that stores data outside of the current Python process. I don't know if that makes sense at all (given that the external support for `bytecode_cache` already exists), and if it's something we want to support. If not, I don't understand why `env.cache` is a public attribute and not preceded by a underscore.

pgjones commented:
@untitaker Thanks; I think that it doesn't make sense to change the `env.cache` given the `bytecode_cache`. If it does however, a potential issue could be the change from str key to tuple, which I made following a [suggestion](https://github.com/pallets/jinja/pull/583#discussion-diff-62604115R772). 

I'd like to add that the performance regression #485 does break jinja for us, so I'd like to help reach a solution.

untitaker commented:
@mitsuhiko confirmed in IRC that cache should not be overridden in such a way.

However, he voiced the concern that a reference to the loader is now always kept. He suggests replacing `loader` with `id(loader)` for the cache key.

pgjones commented:
@untitaker Thanks, change made.

untitaker commented:
Thanks!

pgjones commented:
Thank you

iceboy-sjtu commented:
It seems that cache key should be some weak reference to the loader. By using `id()` you have the potential risk that a new cache having a same address gets old content from the cache.

untitaker commented:
@iceboy-sjtu You may be right, please open a PR.

pgjones commented:
@iceboy-sjtu what about going back to the `hash()`?

iceboy-sjtu commented:
@pgjones `weakref.ref()` seems most suitable for this case. And it's even better to remove the cache entries when a loader is destroyed.

Personally I hate using `hash()` this way, where there will be one in a million case that the hash collides and some mysteries strange behavior appears.
